src/go/rdctl/pkg/snapshot/manager.go

```diff
@@ -96,41 +95,71 @@ func (manager Manager) ValidateName(name string) error {
 	}
 	for _, currentSnapshot := range currentSnapshots {
 		if currentSnapshot.Name == name {
-			return fmt.Errorf("invalid name %q: %w", name, ErrNameExists)
+			return fmt.Errorf("name %q already exists", name)
 		}
 	}
 	return nil
 }
 
+func (manager *Manager) WriteMetadataFile(snapshot Snapshot) (err error) {
+	snapshotDir := manager.SnapshotDirectory(snapshot)
+	if err = os.MkdirAll(snapshotDir, 0o755); err != nil {
+		return fmt.Errorf("failed to create snapshot directory: %w", err)
+	}
+	metadataPath := filepath.Join(snapshotDir, "metadata.json")
+	metadataFile, err := os.Create(metadataPath)
+	if err != nil {
+		return fmt.Errorf("failed to create metadata file: %w", err)
+	}
+	defer metadataFile.Close()
+	encoder := json.NewEncoder(metadataFile)
+	encoder.SetIndent("", "  ")
+	if err = encoder.Encode(snapshot); err != nil {
+		return fmt.Errorf("failed to write metadata file: %w", err)
+	}
+	return
+}
+
 // Create a new snapshot.
-func (manager Manager) Create(name, description string) (*Snapshot, error) {
+func (manager *Manager) Create(name, description string) (snapshot Snapshot, err error) {
+	// Report on invalid names before locking and shutting down the backend
+	if err = manager.ValidateName(name); err != nil {
+		return
+	}
 	id, err := uuid.NewRandom()
 	if err != nil {
-		return nil, fmt.Errorf("failed to generate ID for snapshot: %w", err)
+		return snapshot, fmt.Errorf("failed to generate ID for snapshot: %w", err)
 	}
-	snapshot := Snapshot{
+	snapshot = Snapshot{
 		Created:     time.Now(),
 		Name:        name,
 		ID:          id.String(),
 		Description: description,
 	}
-
-	// do operations that can fail, rolling back if failure is encountered
-	snapshotDir := filepath.Join(manager.Paths.Snapshots, snapshot.ID)
-	if err := manager.Snapshotter.CreateFiles(snapshot); err != nil {
-		if err2 := os.RemoveAll(snapshotDir); err2 != nil {
-			err = errors.Join(err, fmt.Errorf("failed to delete created snapshot directory: %w", err2))
+	if err = lock.Lock(manager.Paths, "create"); err != nil {
```

Comment:
- A number of different unit tests call `manager.Create()`.
- `manager.Create()` calls `lock.Lock()`.
- `lock.Lock()` shuts down the backend via `ensureBackendStopped()`.

When I got this far, I thought: "the unit tests will shut down the backend in a running instance of RD". But I tested it and this is not the case. Here's why:

- `ensureBackendStopped()` calls `GetConnectionInfo()` to get a `ConnectionInfo` instance.
- `GetConnectionInfo()` reads `~/.local/share/rancher-desktop/rd-engine.json` by default on Linux, but this default value is set inside `DefineGlobalFlags()`, which does not run for unit tests. So the default path for the config file is `""`, which causes a `nil` `err` and `nil` `*ConnectionInfo` to be returned from `GetConnectionInfo()`.
- So, the backend does not get shut down.

Many would say this is fine, but this could mean that unit tests start failing when someone makes an unrelated change. I think this exposes a deeper issue: this PR causes `Manager` to depend on the `client` package without any way of mocking `client` for unit testing.

There are no doubt a few ways to solve this problem, but my preference would be to move the `lock.Lock()` call (and associated `lock.Unlock()` call) out of `manager.Create()` and into `createSnapshot()`. Otherwise I suspect we will be doing hacks or making extensive modifications to this PR.

Same comment for `manager.Restore()`.

Reply:
I now agree that it makes sense to include the backend locking/unlocking as part of `Manager`. What you've done looks good to me :+1: